packages/jest-snapshot/src/utils.js

```diff
@@ -21,6 +21,25 @@ const naturalCompare = require('natural-compare');
 const getSerializers = require('./plugins').getSerializers;
 
 const SNAPSHOT_EXTENSION = 'snap';
+const SNAPSHOT_VERSION = '1';
+const SNAPSHOT_VERSION_REGEXP = /^\/\/ Jest Snapshot v(.+),/;
+const SNAPSHOT_GUIDE_LINK = 'https://goo.gl/fbAQLP';
+
+const writeSnapshotVersion = () =>
+  `// Jest Snapshot v${SNAPSHOT_VERSION}, ${SNAPSHOT_GUIDE_LINK}`;
+
+const validateSnapshotVersion = (snapshotContents: string) => {
+  const versionTest = SNAPSHOT_VERSION_REGEXP.exec(snapshotContents);
+  const version = (versionTest && versionTest[1]) || '0';
+
+  if (version < SNAPSHOT_VERSION) {
+    throw new Error(
+      `Stored snapshot version is outdated.\n` +
+      `Expected: v${SNAPSHOT_VERSION}, but received: v${version}\n` +
+      `Update the snapshot to remove this error.`
+    );
```

Comment:
This is a great error but I think we should aim to make this more informative and absolutely solid in messaging. Here is my recommendation:

* If no snapshot header is not found:
```
Outdated snapshot: No snapshot header found. Jest 19 introduced versioned snapshots to ensure all people on a project are using the same version of Jest. Please update all snapshots during this upgrade of Jest.

More information: jest-19-blogpost (I will update this later)
```

* If snapshot header is lower than current version of Jest:
```
Outdated snapshot: The version of the snapshot file associated with this test is outdated. The snapshot file version ensures that all people on a project are using the same version of Jest. Please update all snapshots during this upgrade of Jest.

Expected: v1
Received: v0

More information: jest-19-blogpost (I will update this later)
```

* If snapshot header is higher than current version of Jest:
```
Outdated Jest version: the version of this snapshot file indicates that this project is meant to be used with a newer version of Jest. The snapshot file version ensures that all people on a project are using the same version of Jest. Please update your version of Jest and re-run the tests.

Expected: v1
Received: v2

More information: jest-19-blogpost (I will update this later)
```

We must also include a warning (I think this was brought up by @gaearon yesterday) for the first two to indicate that local change to tests shall be reverted during this update to avoid saving wrong snapshot state. For the first two, how about:

```
Warning: It is advised to revert any local changes to tests or other code during this upgrade to ensure that no invalid state is stored as a snapshot.
```

What do you think?